### PR TITLE
Display the link to add the shipping address when shipping address is not available

### LINKED ISSUE #8027: Display the link to add the shipping address when the "No location by default" is enabled
There is a setting in WooCommerce to set the default customer location. The Cart block won't show any address if there is no default value.
We need to display the link to add the shipping address when the "No location by default" is enabled:

image

## Screenshots

| Actual | Expected |
|---|---|
|image| image |

### Slack
p1671541330197299-slack-C8X6Q7XQU

### PR DESCRIPTION
**Pull request by tarunvijwani**
With this PR, we display a link to add the shipping address in the shipping calculator on the cart page when the shipping address is unavailable. We're also hiding the option to change the shipping address when default shipping rates are defined. 

Fixes #8027 and  #7730

### Changes in the PR
- Replaced hardcoded notices with text for Cart and Checkout block.
- Fixed the shipping placeholder for the Checkout block when no shipping rates are available.
- Created a function in base/utils to format Shipping address.
- Removed the condition to check the default customer address. Display the "Add shipping address" link only when no address is available.
- Hide shipping calculator if default shipping is defined.

#### Accessibility

- [x] I've tested using only a keyboard (no mouse)
- [x] I've tested using a screen reader
- [ ] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [ ] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

### Screenshots

| Before | After |
| ------ | ----- |
| image  |image|

#### Automated Tests
* [x] Changes in this PR are covered by Automated Tests.
  * [x] Unit tests
  * [ ] E2E tests

### case 1 (With no default shipping rates)

image

1. Check out this branch.
2. Go to the WooCommerce settings page: `wp-admin/admin.php?page=wc-settings`.
3. Change 'Default customer location' to `No location by default`. 
4. Remove any default(fallback) shipping rates. 
5. Open your site in a private session(incognito mode).
6. Add some products to the cart. 
7. Go to the cart page, and confirm the "Add an address for shipping rates" link is present in the shipping calculator.
8. Click on the link, and confirm it is working as expected.
9. Open your site in a regular session where you're already logged in and have placed an order before. 
10. Add some products to the cart. 
11. Go to the cart page, confirm the shipping address, and "change address" link present in the shipping calculator.
12. Click on the link, and confirm it is working as expected.

### case 2(With default shipping rates)

image

1. Check out this branch.
2. Go to the WooCommerce settings page: `wp-admin/admin.php?page=wc-settings`.
3. Change 'Default customer location' to `No location by default`. 
4. Add any default(fallback) shipping rates. 
5. Open your site in a private session(incognito mode).
6. Add some products to the cart. 
7. Go to the cart page, and confirm the shipping calculator is hidden and default shipping rates are visible
8. Open your site in a regular session where you're already logged in and have placed an order before. 
9. Add some products to the cart. 
10. Go to the cart page, confirm the shipping address, and "change address" link present in the shipping calculator.
11. Click on the link, and confirm it is working as expected.

### case 3(Error Notice)
1. Check out this branch.
2. Go to the WooCommerce settings page: `wp-admin/admin.php?page=wc-settings`.
3. Change 'Default customer location' to `No location by default`. 
4. Remove any default(fallback) shipping rates. 
5. Open your site in a private session(incognito mode).
6. Add some products to the cart. 
7. Go to the cart page, and confirm the "Add an address for shipping rates" link is present in the shipping calculator.
8. Click on the link, and change the address to a location where shipping rates are unavailable.
9. Confirm that `There are no shipping options available. Please check your shipping address.` error is visible in the text form.
10. Proceed to the Checkout block page, and confirm the `There are no shipping options available. Please check your shipping address` error is visible in the text form under Shipping options. 

* [ ] Do not include in the Testing Notes 

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Changelog

> Display the link to add the shipping address when the shipping address is not available.

### CODE REVIEW

**Comment by mikejolley** [comment]
Before looking at the code, is it not strange that the text here states "Add an address for shipping rates", but still displays shipping rates below? It seems like this link should only change if there are no rates displayed. Unless of course the link text is made slightly more generic so it fits both cases.

**Comment by tarunvijwani** [comment]
>Before looking at the code, is it not strange that the text here states "Add an address for shipping rates", but still displays shipping rates below? It seems like this link should only change if there are no rates displayed. Unless of course the link text is made slightly more generic so it fits both cases.

That's a great point. I think we can change the text to something like "Add an address to get the exact shipping rates"

@elizaan36 @ralucaStan I'd love to hear your thoughts.

File: assets/js/base/components/cart-checkout/totals/shipping/index.tsx
Comment (mikejolley) [inline assets/js/base/components/cart-checkout/totals/shipping/index.tsx]: I notice that in the cart block, if you have an address (whether you've entered it or not) the `ShippingLocation` component is passed a `shippingAddress`. The check inside that component generates a formatted address and returns if null:

```
if ( ! formattedLocation ) {
		return null;
	}
```

Could we instead move that code to a util and consume in both places?

File: src/BlockTypes/Cart.php
Comment (mikejolley) [inline src/BlockTypes/Cart.php]: 👍🏻

**Review by mikejolley — CHANGES_REQUESTED** [review]
The logic makes sense, just a question on reuse and then perhaps a change to the text if needed.

File: assets/js/base/components/cart-checkout/totals/shipping/index.tsx
Comment (tarunvijwani) [reply assets/js/base/components/cart-checkout/totals/shipping/index.tsx]: That's a great suggestion. Thank you!  I have updated the code. 👍

File: assets/js/base/components/cart-checkout/totals/shipping/index.tsx
Comment (tarunvijwani) [inline assets/js/base/components/cart-checkout/totals/shipping/index.tsx]: @mikejolley Do you think we should remove `hasDefaultCustomerAddress` check as well? I think `formatLocation` null check is enough to determine whether we have the address or not to change the label, irrespective of the user's setting in WC. Let me know your thoughts.

Comment (mikejolley) [reply assets/js/base/components/cart-checkout/totals/shipping/index.tsx]: Yeah that makes sense, I'm assuming that formatted location logic needs the basics (state, country, city) or will null. If thats the case, yes remove it.

**Comment by elizaan36** [comment]
hey @tarunvijwani @mikejolley 

It's my understanding that the text "Add an address for shipping rates" would only appear when there are no rates shown. What's the use case where the link would show alongside rates?

**Comment by tarunvijwani** [comment]
> It's my understanding that the text "Add an address for shipping rates" would only appear when there are no rates shown. What's the use case where the link would show alongside rates?

@elizaan36 If the merchant has defined the default shipping rates that are applicable for every location, then it will show the rates.  If the customer adds their address, the shipping rates will change according to their location. The Cart Shortcode page also follows the same approach. 


https://user-images.githubusercontent.com/11503784/213396368-416cf45b-c9e2-4585-afbd-21d1c22e4773.mov





However, if the merchant has not defined the default shipping rates, then it will display the error(We're adding the changes of #8130 to this PR to keep the text consistent) 


https://user-images.githubusercontent.com/11503784/213396404-a9208794-56e8-4db5-8625-7fe1b83c8afc.mov

**Comment by elizaan36** [comment]
> @elizaan36 If the merchant has defined the default shipping rates that are applicable for every location, then it will show the rates.

I imagine the merchant's intention would be to offer the flat rate and turn off the shipping calculator in that case. So I'd say if the merchant has chosen a flat rate for all locations, the link shouldn't appear.

It's quite confusing for the shopper to see "Flat rate" or "Free shipping" with a link that might change these rates. What do you think?

> However, if the merchant has not defined the default shipping rates, then it will display the error(We're adding the changes of https://github.com/woocommerce/woocommerce-blocks/pull/8130 to this PR to keep the text consistent)

Nice, this makes sense and glad we can align the experience there!

**Comment by tarunvijwani** [comment]
>I imagine the merchant's intention would be to offer the flat rate and turn off the shipping calculator in that case. So I'd say if the merchant has chosen a flat rate for all locations, the link shouldn't appear.
>It's quite confusing for the shopper to see "Flat rate" or "Free shipping" with a link that might change these rates. What do you think?

Yes, that makes perfect sense! I'll incorporate the changes. Thank you!

File: assets/js/base/components/cart-checkout/shipping-location/index.tsx
Comment (mikejolley) [inline assets/js/base/components/cart-checkout/shipping-location/index.tsx]: You could move this comment to the interface above and remove the docblock comments here.

File: assets/js/base/components/cart-checkout/totals/shipping/shipping-placeholder.tsx
Comment (mikejolley) [inline assets/js/base/components/cart-checkout/totals/shipping/shipping-placeholder.tsx]: This uses the term "shipping rates", yet the message below and in other places uses "shipping options". We should be consistent because they both mean the same thing.

File: assets/js/base/utils/address.ts
Comment (mikejolley) [inline assets/js/base/utils/address.ts:112]: 👍🏻

File: assets/js/blocks/checkout/inner-blocks/checkout-order-summary-shipping/block.tsx
Comment (mikejolley) [inline assets/js/blocks/checkout/inner-blocks/checkout-order-summary-shipping/block.tsx:29]: What was this for?

**Review by mikejolley — CHANGES_REQUESTED** [review]
Something is broken here. I just tested with no default customer location in a new session, and this is what I got:

Screenshot 2023-01-24 at 11 40 14

No location is shown, and no shipping calculator is shown.

File: assets/js/base/components/cart-checkout/shipping-location/index.tsx
Comment (opr) [inline assets/js/base/components/cart-checkout/shipping-location/index.tsx]: We could remove this, the TS type is good enough I think.

File: assets/js/base/utils/address.ts
Comment (opr) [inline assets/js/base/utils/address.ts]: instead of using `as Record< string, string >` you can pass the generic type to `getSetting` like so:

```suggestion
	const shippingCountries = getSetting< Record< string, string > >(
		'shippingCountries',
		{}
	);
```

This can apply to the other instances of this too :)

File: assets/js/base/components/cart-checkout/totals/shipping/shipping-address.tsx
Comment (opr) [inline assets/js/base/components/cart-checkout/totals/shipping/shipping-address.tsx:41]: Nice change, this will be helpful for #7997

**Review by opr — CHANGES_REQUESTED** [review]
Hi Tarun, thanks for working on this.

It's looking good, but there are a few things that should be looked at:

I see the 'Add an address for shipping rates' displayed inline, I don't see this in the designs, is this intended? I don't think it looks bad, just want to double check.

image

A couple of TS changes, and I won't repeat what Mike already mentioned.

File: assets/js/blocks/checkout/inner-blocks/checkout-order-summary-shipping/block.tsx
Comment (tarunvijwani) [reply assets/js/blocks/checkout/inner-blocks/checkout-order-summary-shipping/block.tsx:29]: I found a bug on the checkout block page that displays `Calculated during checkout` for shipping in the Checkout summary if shipping options are unavailable. It happened because we were not passing the `isCheckout` prop to [TotalsShipping component](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/blocks/checkout/inner-blocks/checkout-order-summary-shipping/block.tsx#L24) for Checkout.  This change will fix it and display the correct error message.
image

**Comment by tarunvijwani** [comment]
> Something is broken here. I just tested with no default customer location in a new session, and this is what I got:
> 
> Screenshot 2023-01-24 at 11 40 14
> 
> No location is shown, and no shipping calculator is shown.

>

@mikejolley do you have default shipping rates enabled? If yes, this is expected as per Test case 2(With default shipping rates). We [decided to remove the shipping calculator](https://github.com/woocommerce/woocommerce-blocks/pull/8141#issuecomment-1397239622) if default shipping rates are defined.

**Comment by tarunvijwani** [comment]
> I see the 'Add an address for shipping rates' displayed inline, I don't see this in the designs, is this intended? I don't think it looks bad, just want to double check.

@opr This might be happening because of the existing CSS for the calculate button component. I changed the text as per Mike's suggestion. Since the text is long so, it appears in the next line now. However, it will vary from theme to theme. I think it is fine to keep the text as is. However, if you feel that we should change the style of the component, let me know. 
image

**Comment by opr** [comment]
Hey @tarunvijwani thanks for explaining, maybe @elizaan36 could help clarify whether this change is ok or if it should be on the line below at all times.

File: assets/js/base/utils/address.ts
Comment (tarunvijwani) [reply assets/js/base/utils/address.ts]: Thank you for the suggestion. I have updated the code.  👍

File: assets/js/base/components/cart-checkout/totals/shipping/shipping-placeholder.tsx
Comment (tarunvijwani) [reply assets/js/base/components/cart-checkout/totals/shipping/shipping-placeholder.tsx]: Right, we should keep the text the same everywhere. I have updated the label. Thank you! 👍

**Comment by tarunvijwani** [comment]
Translation for the text: `Add an address for shipping options` is [not available at the moment](https://translate.wordpress.org/projects/wp-plugins/woo-gutenberg-products-block/stable/nl/default/?filters%5Bterm%5D=Add+an+address+for+shipping+options&filters%5Bterm_scope%5D=scope_any&filters%5Bstatus%5D=current_or_waiting_or_fuzzy_or_untranslated&filters%5Buser_login%5D=&filter=Apply+Filters&sort%5Bby%5D=priority&sort%5Bhow%5D=desc). If the PR is approved, I'll wait till Feb 3 to merge the PR to trunk so that it will be picked in the next WC core release.

**Comment by elizaan36** [comment]
hey thanks for the ping here, Ideally the "add address" link should appear on it's own line (that was the intention) but it's not a huge UX issue if it appears inline in some cases. 

Looking at the screenshot now I'm mainly curious why it's not picking up the same text links styles as the rest of the page? The text is gray and heavier weight, which creates a lot of visual noise. It should pull in the accent color defined by the theme for text links and match the font weight.

image

**Comment by tarunvijwani** [comment]
@elizaan36 Fixed it in the [commit](https://github.com/woocommerce/woocommerce-blocks/pull/8141/commits/0b0d6186cb874acc7a039a1eb05aaf042acb5d90):

image

File: assets/js/base/components/cart-checkout/totals/shipping/calculator-button.tsx
Comment (opr) [inline assets/js/base/components/cart-checkout/totals/shipping/calculator-button.tsx:20]: This causes the URL in the browser to update when clicking. It is unclear to me why we have to show a link instead of a button when rates are not available, what is the specific reason for showing different elements?

To avoid the URL change you can pass the event as a parameter of the `onClick` handler, and call `.preventDefault()` on it, like so:

```js
onClick={ ( e ) => {
  e.preventDefault();
  setIsShippingCalculatorOpen( ! isShippingCalculatorOpen );
} }
```

**Review by opr — COMMENTED** [review]
Hey Tarun, when I add an address that does not have rates I see this:

image

I wonder if it would be better if we checked to see if the address is not empty, and if so, we could change the text to `Add a different address for shipping options`. The reason this may be better is because the shopper could be confused as they would have just added an address, so why would the page be asking them to add one? Also added a comment about the link/button situation, it would be great if you could provide more info on that! 🙇🏼

**Comment by tarunvijwani** [comment]
> I wonder if it would be better if we checked to see if the address is not empty, and if so, we could change the text to Add a different address for shipping options. The reason this may be better is because the shopper could be confused as they would have just added an address, so why would the page be asking them to add one? Also added a comment about the link/button situation, it would be great if you could provide more info on that! 🙇🏼

That's a great suggestion. However, if we change the text, it will be misaligned for some themes:

image


I tried adding the `white-space: break-spaces;` in CSS but it is looking ugly. 


We can try a shorter length string like `Add a different shipping address`. Let me know your thoughts.

File: assets/js/base/components/cart-checkout/totals/shipping/calculator-button.tsx
Comment (tarunvijwani) [reply assets/js/base/components/cart-checkout/totals/shipping/calculator-button.tsx:20]: Ah, I missed adding the ` e.preventDefault();`. Thank you, I have updated the code.  

>It is unclear to me why we have to show a link instead of a button when rates are not available, what is the specific reason for showing different elements?

It's because of the styling of the elements. I kept the original button for the conditions where we are displaying the `calculate` link. 
For others, I'm displaying the `Add an address` link to match the style with the coupon code link. 

I tried using the link at both places but then the `calculate` link gets the theme color which was taking too much attention. So I decided to keep the original style and elements for it. Please let me know if there is a better way to tackle it.

**Review by opr — COMMENTED** [review]
Ok, thanks for the answer, if we want to keep it in line with the coupon link this is OK 👍🏼 

As for the text not breaking, this is caused by https://github.com/woocommerce/woocommerce-blocks/blob/669aee78ad4b6c97388b2e75e444ebb38df23be5/packages/checkout/components/totals/item/style.scss#L14 so maybe we can apply a new class, or modifier or something and remove this specific rule for the add/update shipping address link?

**Comment by elizaan36** [comment]
> I wonder if it would be better if we checked to see if the address is not empty, and if so, we could change the text to Add a different address for shipping options. The reason this may be better is because the shopper could be confused as they would have just added an address, so why would the page be asking them to add one?

@tarunvijwani @opr Would it be possible to align with the behavior on the checkout block for this use case. 
- User enters an address that is not within any shipping zone
- Notice appears "There are no shipping options available. Please check your shipping address."

**Comment by elizaan36** [comment]
Something like this - thoughts?
image

**Comment by tarunvijwani** [comment]
@elizaan36 Just wanted to share the final output of the PR. Please feel free to add any changes and suggestions. Thank you! 

## Case 1: With no default shipping method:

https://user-images.githubusercontent.com/11503784/220379739-8f634652-0fbd-4dbf-bd6c-a329778c58b0.mov

## Case 2: With default shipping method:

https://user-images.githubusercontent.com/11503784/220380918-5f7a9ab6-2da4-4835-bfc6-9d642b0519f3.mov

**Comment by elizaan36** [comment]
Thanks for sharing @tarunvijwani ! One small thing. To help with visual rhythm and scanning the page, can we match the padding top and bottom around the rates? Here's a sketch to illustrate. 
Screen Shot 2023-02-21 at 9 25 02 AM

The checkout block has a good amount of padding around the rates, so we should aim for a cohesive look and feel on the Cart.

**Comment by tarunvijwani** [comment]
> One small thing. To help with visual rhythm and scanning the page, can we match the padding top and bottom around the rates? Here's a sketch to illustrate.

Thank you for bringing it up. I have updated the margin now:

| Twenty Twenty-Three | Storefront |
|---|---|
|image|image|

**Comment by ralucaStan** [comment]
Noting here as well that we don't have the informational notices (that @elizaan36 used in the design) implemented yet in the project.

So we can either use the text variant from Tarun's videos or we can show the error notice and follow up with some other tickets in the future to replace it with the information one.

**Comment by tarunvijwani** [comment]
Converted the PR to draft to add unit test cases.

**Comment by tarunvijwani** [comment]
@opr While creating the unit test cases, I found an error that `formatShippingAddress` returns the formatted address string even if the postcode or any other field is missing.  I have created the `isAddressComplete` function to validate the address and replaced the `formatShippingAddress` function with it. Please let me know if you have any questions.

File: assets/js/base/utils/test/address.ts
Comment (opr) [inline assets/js/base/utils/test/address.ts]:
```suggestion
	it( 'correctly checks complete addresses', () => {
```

File: assets/js/base/utils/test/address.ts
Comment (opr) [inline assets/js/base/utils/test/address.ts]:
```suggestion
	it( 'correctly checks incomplete addresses', () => {
```

File: assets/js/base/utils/test/address.ts
Comment (opr) [inline assets/js/base/utils/test/address.ts]:
```suggestion
	it( 'correctly checks empty addresses', () => {
```

File: assets/js/base/utils/test/address.ts
Comment (opr) [inline assets/js/base/utils/test/address.ts]:
```suggestion
	it( 'returns null if address is empty', () => {
```

File: assets/js/base/components/cart-checkout/totals/shipping/shipping-address.tsx
Comment (opr) [inline assets/js/base/components/cart-checkout/totals/shipping/shipping-address.tsx]:
```suggestion
	// If the address is incomplete, and we're not in the editor, don't show anything.
```

File: assets/js/base/utils/address.ts
Comment (opr) [inline assets/js/base/utils/address.ts]:
```suggestion
 * Returns true if the address has a city, country, and postcode.
```

File: assets/js/base/utils/address.ts
Comment (opr) [inline assets/js/base/utils/address.ts]: Nitpick, sorry, but comments should begin with a capital letter.

```suggestion
	// We bail early if we don't have an address.
```

**Review by opr — CHANGES_REQUESTED** [review]
The new tests look nice, I suggested some improvements to the names to make them a little more accurate. By the way there is a typo in the email and customer name, it should be `John Doe` (this isn't technically important as the tests still work, but it is just a small detail that makes our repo appear more polished).

Besides this, I still really think we should change the wording when an address has _already_ been entered.

image

If the user is being asked to add an address, it implies one has not been entered. We can use your `isAddressComplete` function combined with the number of rates to determine what to show in this case.

If this presents too many problems, let's resolve the current suggestions, and we can merge this and address the link wording in a follow-up PR, how does that sound?

File: assets/js/base/utils/address.ts
Comment (tarunvijwani) [inline assets/js/base/utils/address.ts:162]: Some countries like UAE don't have postcodes. The `isAddressComplete` function was returning `false` for these countries. I have now removed the postcode check from the `isAddressComplete` function to ensure it works for all countries. We already have the postcode validation in the shipping calculator that works for the countries which require postcodes.

**Review by opr — APPROVED** [review]
Hey @tarunvijwani nice it looks good, sorry I missed the change with the address shown!

**Comment by tarunvijwani** [comment]
We have open #8621 PR to add more unit tests to this one. Once #8621 is merged, we will test this PR again and merge it.

**Comment by tarunvijwani** [comment]
This PR will be merged in the 1st week of April and released in the 10.0.0 version.

File: assets/js/base/components/cart-checkout/totals/shipping/index.tsx
Comment (tarhi-saad) [inline assets/js/base/components/cart-checkout/totals/shipping/index.tsx:87]: @tarunvijwani, on the Checkout Block page, we would like to show the "ShippingPlaceholder" component when there are no shipping methods available (i.e., `hasRates && cartHasCalculatedShipping`). But, with this additional condition (i.e., `! addressComplete &&`), the Shipping placeholder will never be displayed if the address is complete!

**Review by tarhi-saad — COMMENTED** [review]
@tarunvijwani, there is a logical error in the code! Please check my comment below!

File: assets/js/base/components/cart-checkout/totals/shipping/index.tsx
Comment (tarunvijwani) [reply assets/js/base/components/cart-checkout/totals/shipping/index.tsx:87]: Continuing the discussion on slack: p1679930401191499-slack-C8X6Q7XQU

**Pull request merged**
